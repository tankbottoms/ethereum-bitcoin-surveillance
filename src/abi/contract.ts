/* eslint-disable eqeqeq */
import { EthereumClient } from '../eth/client';
import { getCode } from '../eth/requests';
import { Address } from '../core/msgs';
import { createModuleDebug, TRACE_ENABLED } from '../utils/debug';
import { sha3 } from './wasm';

const { debug, trace } = createModuleDebug('abi:contract');

export interface ContractInfo {
    /** True if the corresponding account is a smart contract, otherwise false */
    isContract: boolean;
    /** A unique representation of all function and event signatures of a contract */
    fingerprint?: string;
    /** Name of the contract from configured ABIs */
    contractName?: string;
}

/** Lookup function to find matching signature for a hash */
export type SignatureMatcher = (sig: string, address?: Address) => string | undefined;

/** Lookup function to find name for a given contract address or fingerprint */
export type ContractNameLookup = (address: Address, fingerprint: string) => string | undefined;

/**
 * Find function and event signature hashes in the EVM bytecode and attempt to match them against
 * known hashes from configured ABIs
 */
export function extractFunctionsAndEvents(
    contractCode: string,
    signatureMatcher: SignatureMatcher
): { functions: string[]; events: string[] } | undefined {
    const functionMatches: Set<string> = new Set();
    const eventMatches: Set<string> = new Set();
    const code = Buffer.from(contractCode.slice(2), 'hex');
    for (let i = 0, len = code.length; i < len; i++) {
        const opcode = code[i];
        // Look for PUSH<n> opcodes - their pushData may contain a log topic or function signature hash
        if (opcode >= 0x60 && opcode <= 0x7f) {
            const dataLength = opcode - 0x5f;
            if (dataLength === 32 || dataLength === 4) {
                const data = code.slice(i + 1, i + dataLength + 1).toString('hex');
                const match = signatureMatcher(data);
                if (match) {
                    (dataLength === 4 ? functionMatches : eventMatches).add(match);
                }
            }
            i += dataLength;
        }
    }

    const functions = [...functionMatches].sort();
    const events = [...eventMatches].sort();
    return { functions, events };
}

/** Creates a hash of all function and event signatures of a contract used to match ABIs against EVM bytecode */
export function computeContractFingerprint(
    { functions, events }: { functions: string[]; events: string[] } = { functions: [], events: [] }
): string | undefined {
    if (functions.length === 0 && events.length === 0) {
        return;
    }
    const fingerprint = sha3(`${functions.join(',')}|${events.join(',')}`)!.slice(2);
    return fingerprint;
}

/**
 * Attempts to determine if the given account address is a smart contract and
 * determines some additional information if that's the case
 */
export async function getContractInfo(
    address: Address,
    ethClient: EthereumClient,
    signatureMatcher?: SignatureMatcher,
    contractNameLookup?: ContractNameLookup
): Promise<ContractInfo> {
    debug('Retrieving contract information for address %s', address);
    const code = await ethClient.request(getCode(address));
    if (code === '0x') {
        return { isContract: false };
    }
    if (signatureMatcher == null) {
        return { isContract: true };
    }
    const fnsEvts = extractFunctionsAndEvents(code, (_fingerprint: string) => signatureMatcher(_fingerprint, address));
    const fingerprint = computeContractFingerprint(fnsEvts);
    if (TRACE_ENABLED) {
        trace('Computed fingerprint %s from contract code %O', fingerprint, fnsEvts);
    }
    const contractName =
        fingerprint != null && contractNameLookup != null ? contractNameLookup(address, fingerprint) : undefined;
    return {
        isContract: true,
        fingerprint,
        contractName,
    };
}
