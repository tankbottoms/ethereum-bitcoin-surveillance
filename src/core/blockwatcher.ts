/* eslint-disable eqeqeq */
import { ContractInfo, getContractInfo } from '../abi/contract';
import { AbiRepository } from '../abi/repo';
import { BlockRange, blockRangeSize, blockRangeToArray, chunkedBlockRanges, serializeBlockRange } from './blockrange';
import { Checkpoint } from './state';
import { BlockWatcherConfig } from './config';
import { EthereumClient } from '../eth/client';
import { blockNumber, getBlock, getTransactionReceipt } from '../eth/requests';
import { RawBlockResponse, RawLogResponse, RawTransactionReceipt, RawTransactionResponse } from '../eth/responses';
import { formatBlock, formatLogEvent, formatTransaction } from './format';
import { Address, AddressInfo, FormattedBlock, LogEventMessage, PrivateTransactionPayload } from './msgs';
import { Output, OutputMessage } from './output';
import { isEnterprisePlatform, NodePlatformAdapter } from '../platforms';
import { ABORT, AbortHandle } from '../utils/abort';
import { parallel, sleep } from '../utils/async';
import { bigIntToNumber } from '../utils/bn';
import { Cache, cachedAsync, NoopCache } from '../utils/cache';
import { createModuleDebug } from '../utils/debug';
import { ManagedResource } from '../utils/resource';
import { linearBackoff, resolveWaitTime, retry, WaitTime } from '../utils/retry';
import { AggregateMetric } from '../utils/stats';

const { debug, info, warn, error, trace } = createModuleDebug('blockwatcher');

export type StartBlock = 'latest' | 'genesis' | number;

const toAddressInfo = (contractInfo?: ContractInfo): AddressInfo | undefined =>
  contractInfo
    ? {
        contractName: contractInfo.contractName,
        isContract: contractInfo.isContract,
      }
    : undefined;

export function parseBlockTime(timestamp: number | string): number {
  if (typeof timestamp === 'number') {
    return timestamp * 1000;
  }

  if (typeof timestamp === 'string') {
    // Timestamp on Quorum/Raft nodes are emitted as nanoseconds since unix epoch
    // so translate it to milliseconds
    const ms = BigInt(timestamp) / BigInt(1_000_000);
    return bigIntToNumber(ms);
  }
  throw new Error(`Unable to parse block timestamp "${timestamp}"`);
}

const initialCounters = {
  blocksProcessed: 0,
  transactionsProcessed: 0,
  transactionLogEventsProcessed: 0,
};

/**
 * BlockWatcher will query all blocks, transactions, receipts and logs from an ethereum node,
 * compute some additional information, such as address information and decoded ABI data and
 * will send them to the output. It uses checkpoints to keep track of state and will resume
 * where it left off when restarted.
 */
export class BlockWatcher implements ManagedResource {
  private active: boolean = true;
  private ethClient: EthereumClient;
  private checkpoint: Checkpoint;
  private output: Output;
  private abiRepo?: AbiRepository;
  private config: BlockWatcherConfig;
  private nodePlatform: NodePlatformAdapter;
  private abortHandle = new AbortHandle();
  private endCallbacks: Array<() => void> = [];
  private waitAfterFailure: WaitTime;
  private chunkQueueMaxSize: number;
  private contractInfoCache: Cache<string, Promise<ContractInfo>> = new NoopCache();
  private counters = { ...initialCounters };
  private aggregates = {
    blockProcessTime: new AggregateMetric(),
    txProcessTime: new AggregateMetric(),
    eventProcessTime: new AggregateMetric(),
  };

  constructor({
    ethClient,
    checkpoint,
    output,
    abiRepo,
    contractInfoCache,
    config,
    waitAfterFailure = linearBackoff({ min: 0, step: 2500, max: 120_000 }),
    chunkQueueMaxSize = 1000,
    nodePlatform,
  }: {
    ethClient: EthereumClient;
    checkpoint: Checkpoint;
    output: Output;
    abiRepo: AbiRepository;
    config: BlockWatcherConfig;
    waitAfterFailure?: WaitTime;
    chunkQueueMaxSize?: number;
    contractInfoCache?: Cache<string, Promise<ContractInfo>>;
    nodePlatform: NodePlatformAdapter;
  }) {
    this.ethClient = ethClient;
    this.checkpoint = checkpoint;
    this.output = output;
    this.abiRepo = abiRepo;
    this.config = config;
    this.waitAfterFailure = waitAfterFailure;
    this.chunkQueueMaxSize = chunkQueueMaxSize;
    this.nodePlatform = nodePlatform;
    if (contractInfoCache) {
      this.contractInfoCache = contractInfoCache;
    }
  }

  async start(): Promise<void> {
    debug('Starting block watcher');
    const { ethClient, checkpoint } = this;

    if (checkpoint.isEmpty()) {
      if (typeof this.config.startAt === 'number') {
        if (this.config.startAt < 0) {
          const latestBlockNumber = await ethClient.request(blockNumber());
          checkpoint.setInitialBlockNumber(Math.max(0, latestBlockNumber + this.config.startAt));
        } else {
          info(`setting numeric start block number: ${this.config.startAt}`);
          checkpoint.setInitialBlockNumber(this.config.startAt);
        }
      } else if (this.config.startAt === 'genesis') {
        checkpoint.setInitialBlockNumber(0);
      } else if (this.config.startAt === 'latest') {
        const latestBlockNumber = await ethClient.request(blockNumber());
        checkpoint.setInitialBlockNumber(latestBlockNumber);
      } else {
        throw new Error(`Invalid start block: ${JSON.stringify(this.config.startAt)}`);
      }
      info(
        'Determined initial block number: %d from configured value %o',
        checkpoint.getinitialBlockNumber(),
        this.config.startAt,
      );
    }
    let failures = 0;
    while (this.active) {
      try {
        const latestBlockNumber = await ethClient.request(blockNumber());
        debug('Received latest block number: %d', latestBlockNumber);
        const todo = checkpoint.getIncompleteRanges(latestBlockNumber);
        debug('Found %d block ranges to process', todo.length);
        if (todo.length) {
          info(
            'Latest block number is %d - processing remaining %d blocks',
            latestBlockNumber,
            todo.map(blockRangeSize).reduce((a, b) => a + b, 0),
          );
        }

        for (const range of todo) {
          if (!this.active) {
            throw ABORT;
          }
          await parallel(
            chunkedBlockRanges(range, this.config.blocksMaxChunkSize, this.chunkQueueMaxSize).map(chunk => async () => {
              if (!this.active) {
                return;
              }
              await retry(() => this.processChunk(chunk), {
                attempts: 100,
                waitBetween: this.waitAfterFailure,
                taskName: `block range ${serializeBlockRange(chunk)}`,
                abortHandle: this.abortHandle,
                warnOnError: true,
                onRetry: attempt =>
                  warn('Retrying to process block range %s (attempt %d)', serializeBlockRange(chunk), attempt),
              });
            }),
            { maxConcurrent: this.config.maxParallelChunks, abortHandle: this.abortHandle },
          );
          failures = 0;
        }
      } catch (e) {
        if (e === ABORT) {
          break;
        }
        error('Error in block watcher polling loop', e);
        failures++;
        const waitTime = resolveWaitTime(this.waitAfterFailure, failures);
        if (waitTime > 0) {
          warn('Waiting for %d ms after %d consecutive failures', waitTime, failures);
          await this.abortHandle.race(sleep(waitTime));
        }
      }
      if (this.active) {
        try {
          await this.abortHandle.race(sleep(this.config.pollInterval));
        } catch (e) {
          if (e === ABORT) {
            break;
          } else {
            warn('Unexpected error while waiting for next polling loop iteration', e);
          }
        }
      }
    }
    if (this.endCallbacks != null) {
      this.endCallbacks.forEach(cb => cb());
    }
    info('Block watcher stopped');
  }

  async processChunk(chunk: BlockRange) {
    const startTime = Date.now();
    info('Processing chunk %s', serializeBlockRange(chunk));

    debug('Requesting block range', chunk);
    const blockRequestStart = Date.now();
    const blocks = await this.ethClient
      .requestBatch(
        blockRangeToArray(chunk)
          .filter(blockNum => this.checkpoint.isIncomplete(blockNum))
          .map(blockNum => getBlock(blockNum)),
      )
      .catch(e => Promise.reject(new Error(`Failed to request batch of blocks ${serializeBlockRange(chunk)}: ${e}`)));
    debug('Received %d blocks in %d ms', blocks.length, Date.now() - blockRequestStart);
    for (const block of blocks) {
      await this.processBlock(block);
    }
    info('Completed %d blocks of chunk %s in %d ms', blocks.length, serializeBlockRange(chunk), Date.now() - startTime);
  }

  async processBlock(block: RawBlockResponse) {
    if (block.number != null && !this.checkpoint.isIncomplete(bigIntToNumber(block.number))) {
      warn('Skipping processing of block %d since it is marked complete in our checkpoint');
      return;
    }
    const startTime = Date.now();
    const outputMessages: OutputMessage[] = [];
    const formattedBlock = formatBlock(block);
    if (formattedBlock.number == null) {
      debug('Ignoring block %s without number', block.hash);
      return;
    }
    const blockTime = parseBlockTime(formattedBlock.timestamp);
    outputMessages.push({
      type: 'block',
      time: blockTime,
      body: formattedBlock,
    });
    const txMsgs = await this.abortHandle.race(
      Promise.all(block.transactions.map(tx => this.processTransaction(tx, blockTime, formattedBlock))),
    );
    if (!this.active) {
      return;
    }
    outputMessages.forEach(msg => this.output.write(msg));
    txMsgs.forEach(msgs => msgs.forEach(msg => this.output.write(msg)));

    this.checkpoint.markBlockComplete(formattedBlock.number);
    this.counters.blocksProcessed++;
    this.aggregates.blockProcessTime.push(Date.now() - startTime);
  }

  async processTransaction(
    rawTx: RawTransactionResponse | string,
    blockTime: number,
    formattedBlock: FormattedBlock,
  ): Promise<OutputMessage[]> {
    if (!this.active) {
      return [];
    }
    if (typeof rawTx === 'string') {
      warn('Received raw transaction as string from block %d', formattedBlock.number);
      return [];
    }
    const startTime = Date.now();
    trace('Processing transaction %s from block %d', rawTx.hash, formattedBlock.number);

    try {
      let receipt: RawTransactionReceipt | undefined = undefined;

      try {
        receipt = await this.ethClient.request(getTransactionReceipt(rawTx.hash));
      } catch (err) {
        // trace(`No receipt`);
      }
      const [toInfo, fromInfo] = await Promise.all([
        rawTx.to != null ? this.lookupContractInfo(rawTx.to) : undefined,
        rawTx.from != null ? this.lookupContractInfo(rawTx.from) : undefined,
      ]);

      let contractAddresInfo: ContractInfo | undefined;
      if (receipt?.contractAddress != null) {
        contractAddresInfo = await this.lookupContractInfo(receipt.contractAddress);
      }

      let privatePayload: PrivateTransactionPayload | undefined;
      if (this.config.decryptPrivateTransactions) {
        // geth is not an EnterprisePlatform.
        if (isEnterprisePlatform(this.nodePlatform)) {
          if (this.nodePlatform.isPrivateTransaction(rawTx)) {
            info('Retrieving decrypted private transaction input %o', {
              txHash: rawTx.hash,
              input: rawTx.input,
            });
            try {
              const input = await this.nodePlatform.getRawTransactionInput(rawTx.input, this.ethClient);
              info('Successfully retrieved decrypted transaction payload %o', { input });
              privatePayload = { input };
            } catch (e) {
              error(
                'Failed to retrieve decrypted transaction payload %o',
                {
                  txHash: rawTx.hash,
                  input: rawTx.input,
                },
                e,
              );
            }
          }
        } else {
          warn(
            "Decryption of private transactions is enabled, but detected node type (%s) doesn't support it",
            this.nodePlatform.name,
          );
        }
      }

      let callInfo;
      if (this.abiRepo && toInfo && toInfo.isContract) {
        //info('trying to decode fn')
        callInfo = this.abiRepo.decodeFunctionCall(privatePayload?.input ?? rawTx.input, {
          contractAddress: rawTx.to ?? undefined,
          contractFingerprint: toInfo.fingerprint,
        });
      }

      this.counters.transactionsProcessed++;
      this.aggregates.txProcessTime.push(Date.now() - startTime);
      trace('Completed processing transaction %s in %d ms', rawTx.hash, Date.now() - startTime);
      return [
        {
          type: 'transaction',
          time: blockTime,
          body: formatTransaction(
            rawTx,
            receipt,
            toAddressInfo(fromInfo),
            toAddressInfo(toInfo),
            toAddressInfo(contractAddresInfo),
            callInfo,
            privatePayload,
          ),
        },
        ...(await Promise.all(receipt?.logs?.map(l => this.processTransactionLog(l, blockTime)) ?? [])),
      ];
    } catch (err) {
      error(`proc tx ${JSON.stringify(err, null, 2)}`);
      throw err;
    }
  }

  async processTransactionLog(evt: RawLogResponse, blockTime: number): Promise<LogEventMessage> {
    const startTime = Date.now();
    const contractInfo = await this.lookupContractInfo(evt.address);
    const decodedEventData = this.abiRepo?.decodeLogEvent(evt, {
      contractAddress: evt.address,
      contractFingerprint: contractInfo?.fingerprint,
    });
    this.aggregates.eventProcessTime.push(Date.now() - startTime);
    this.counters.transactionLogEventsProcessed++;
    return {
      type: 'event',
      time: blockTime,
      body: formatLogEvent(evt, toAddressInfo(contractInfo), decodedEventData),
    };
  }

  async lookupContractInfo(address: Address): Promise<ContractInfo | undefined> {
    const abiRepo = this.abiRepo;
    if (abiRepo == null) {
      return;
    }
    const result = await cachedAsync(address, this.contractInfoCache, (addr: Address) =>
      getContractInfo(
        addr,
        this.ethClient,
        (sig: string) => abiRepo.getMatchingSignature(sig),
        (_address: string, fingerprint: string) =>
          abiRepo.getContractByAddress(_address)?.contractName ??
          abiRepo.getContractByFingerprint(fingerprint)?.contractName,
      ),
    );
    return result;
  }

  public async shutdown() {
    info('Shutting down block watcher');
    this.active = false;
    this.abortHandle.abort();
    await new Promise<void>(resolve => {
      this.endCallbacks.push(resolve);
    });
  }

  public flushStats() {
    const stats = {
      ...this.counters,
      ...this.aggregates.blockProcessTime.flush('blockProcessTime'),
      ...this.aggregates.txProcessTime.flush('txProcessTime'),
      ...this.aggregates.eventProcessTime.flush('eventProcessTime'),
      abortHandles: this.abortHandle.size,
    };
    this.counters = { ...initialCounters };
    return stats;
  }
}
