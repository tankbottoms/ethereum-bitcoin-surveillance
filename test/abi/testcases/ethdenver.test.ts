import debugModule from 'debug';
import { join } from 'path';
import { ContractInfo } from '../../../src/abi/contract';
import { AbiRepository } from '../../../src/abi/repo';
import { BlockWatcher } from '../../../src/blockwatcher';
import { State } from '../../../src/state';
import { BatchedEthereumClient } from '../../../src/eth/client';
import { HttpTransport } from '../../../src/eth/http';
import { withRecorder } from '../../../src/eth/recorder';
import { MOCK_NODE_ADAPTER } from '../../../src/platforms/mock';
import { enableTraceLogging, suppressDebugLogging } from '../../../src/utils/debug';
import { LRUCache } from '../../../src/utils/lru';
import { TestOutput } from '../../testoutput';

let logHandle: any;
beforeEach(() => {
    logHandle = suppressDebugLogging();
});
afterEach(() => {
    logHandle.restore();
});

test('blockwatcher', async () => {
    const BLOCK = 8394957;
    enableTraceLogging();
    debugModule.enable('ethlogger:abi:*');
    await withRecorder(
        new HttpTransport('https://dai.poa.network', {}),
        {
            name: `testcases-ethdenver-${BLOCK}`,
            storageDir: join(__dirname, '../../fixtures/recorded'),
            replay: true,
        },
        async transport => {
            const ethClient = new BatchedEthereumClient(transport, { maxBatchSize: 100, maxBatchTime: 0 });
            const abiRepo = new AbiRepository({
                decodeAnonymous: true,
                fingerprintContracts: true,
                abiFileExtension: '.json',
                directory: join(__dirname, '../../abis'),
                searchRecursive: true,
                requireContractMatch: true,
                reconcileStructShapeFromTuples: false,
            });
            await abiRepo.initialize();
            const checkpoints = new State({
                path: join(__dirname, '../../../tmp/tmpcheckpoint.json'),
                saveInterval: 10000,
            });
            const checkpoint = checkpoints.getCheckpoint('main');
            checkpoint.setInitialBlockNumber(0);
            const output = new TestOutput();
            const contractInfoCache = new LRUCache<string, Promise<ContractInfo>>({ maxSize: 100 });
            const blockWatcher = new BlockWatcher({
                abiRepo,
                checkpoint,
                config: {
                    enabled: true,
                    blocksMaxChunkSize: 1,
                    pollInterval: 1,
                    maxParallelChunks: 1,
                    startAt: 'latest',
                    decryptPrivateTransactions: false,
                    retryWaitTime: 10,
                },
                ethClient,
                output,
                contractInfoCache,
                waitAfterFailure: 1,
                nodePlatform: MOCK_NODE_ADAPTER,
            });

            await blockWatcher.processChunk({ from: BLOCK, to: BLOCK });

            expect(output.messages).toMatchSnapshot();
        }
    );
}, 15000);
