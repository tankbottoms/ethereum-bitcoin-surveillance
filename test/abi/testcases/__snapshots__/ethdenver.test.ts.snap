// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blockwatcher 1`] = `
Array [
  Object {
    "body": Object {
      "difficulty": "340282366920938463463374607431768211454",
      "extraData": "0xde830206088f5061726974792d457468657265756d86312e34302e30826c69",
      "gasLimit": 10000000,
      "gasUsed": 223842,
      "hash": "0x05a82978c95cf9e30045e0eb9b9cb48d5639b3ab9b8007b2f8d371dc99ed0afd",
      "logsBloom": "0x04000000000000000000000000000000000000000000000000000000000000000000000010000000000008000000000010000010000000000000000000000040000000002040000000000018000000000000000000001000042000000000000000000000000000000400000000000000000000400000000000000010000000000000000000000000000000000000000000100000020000000000000000000000000000000100000000000000000000000000000000400000000000000000000040000022000000002000000000000000000000008000000000002000000000000010000000000000000000000000200000004000000000000000000000000000",
      "miner": "0xA13D45301207711B7C0328c6b2b64862abFe9b7a",
      "nonce": undefined,
      "number": 8394957,
      "parentHash": "0x25a7c22beedc516f2f3a6e8c78589417664e4f278f8fe8875397edd03e590287",
      "receiptsRoot": "0xd1363c7ade1bc9c1b5fb95d1c29c1b153b1ceae9cef5d432d1036e4d2dfbe0fd",
      "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
      "size": 1696,
      "stateRoot": "0x7dc259f4d32d7f39126fdebbcacac79fa2aad890e7b5bead633030c5d5c19ebe",
      "timestamp": 1581720675,
      "totalDifficulty": "2856655838159500800421100904281574568815419372",
      "transactionCount": 1,
      "transactionsRoot": "0x4e7ad1168e367757b722a0df0568d5660811891576e7ea92bba91c169304e140",
      "uncles": Array [],
    },
    "time": 1581720675000,
    "type": "block",
  },
  Object {
    "body": Object {
      "blockHash": "0x05a82978c95cf9e30045e0eb9b9cb48d5639b3ab9b8007b2f8d371dc99ed0afd",
      "blockNumber": 8394957,
      "call": Object {
        "args": undefined,
        "name": "relayCall",
        "params": Array [
          Object {
            "name": undefined,
            "type": "address",
            "value": "0x150042AAe387E5c002dfaF7844b3110828094ACA",
          },
          Object {
            "name": undefined,
            "type": "address",
            "value": "0x6C85E15838034B7BbA6C63bAdD8839fd4c5BE606",
          },
          Object {
            "name": undefined,
            "type": "bytes",
            "value": "0xd80106b900000000000000000000000047643fd12a9464ba6cbadd6d5afc095756640d0b000000000000000000000000bb2692ec9976975a8cb80b52832a94273fbfb43800000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000000849bd9bbc60000000000000000000000000f5ef980513ca22faf70243959c92cf5ee1e531b000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000041fbd8c79438ff87e7cef097e14e5e820f83c637d5a963c470751916afed65148f635c56ba621c554725bf67eca7a06f600641b8fc3e2385671a15810b194baae31b00000000000000000000000000000000000000000000000000000000000000",
          },
          Object {
            "name": undefined,
            "type": "uint256",
            "value": 70,
          },
          Object {
            "name": undefined,
            "type": "uint256",
            "value": 1000000000,
          },
          Object {
            "name": undefined,
            "type": "uint256",
            "value": 1536276,
          },
          Object {
            "name": undefined,
            "type": "uint256",
            "value": 1,
          },
          Object {
            "name": undefined,
            "type": "bytes",
            "value": "0x6fb1c4d50445a989e871840ec68d891a257af43d64d59e29a2ffb4b5a0a1d41d7f79914ef0b8b6697ca3df44ca1e490c1aa3c43f345962858ef34f7c186ce5321c",
          },
          Object {
            "name": undefined,
            "type": "bytes",
            "value": null,
          },
        ],
        "signature": "relayCall(address,address,bytes,uint256,uint256,uint256,uint256,bytes,bytes)",
      },
      "contractAddress": null,
      "contractAddressInfo": undefined,
      "cumulativeGasUsed": 223842,
      "from": "0x9718A022941447fa2aF776eCbde7007bBA5A7edd",
      "fromInfo": Object {
        "contractName": undefined,
        "isContract": false,
      },
      "gas": 1945312,
      "gasPrice": 1000000000,
      "gasUsed": 223842,
      "hash": "0xff36619285cf23ef7f812ea3600a369b52ebd392e682deeb77dd758b4462f38f",
      "input": "0x405cec67000000000000000000000000150042aae387e5c002dfaf7844b3110828094aca0000000000000000000000006c85e15838034b7bba6c63badd8839fd4c5be60600000000000000000000000000000000000000000000000000000000000001200000000000000000000000000000000000000000000000000000000000000046000000000000000000000000000000000000000000000000000000003b9aca0000000000000000000000000000000000000000000000000000000000001771140000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000034000000000000000000000000000000000000000000000000000000000000003c000000000000000000000000000000000000000000000000000000000000001e4d80106b900000000000000000000000047643fd12a9464ba6cbadd6d5afc095756640d0b000000000000000000000000bb2692ec9976975a8cb80b52832a94273fbfb43800000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000000849bd9bbc60000000000000000000000000f5ef980513ca22faf70243959c92cf5ee1e531b000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000041fbd8c79438ff87e7cef097e14e5e820f83c637d5a963c470751916afed65148f635c56ba621c554725bf67eca7a06f600641b8fc3e2385671a15810b194baae31b000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000416fb1c4d50445a989e871840ec68d891a257af43d64d59e29a2ffb4b5a0a1d41d7f79914ef0b8b6697ca3df44ca1e490c1aa3c43f345962858ef34f7c186ce5321c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 556,
      "privatePayload": undefined,
      "r": "0x7843ec8bc029c4f623ece12111abd8d361a96aed84a3851def9bccbec6456ae",
      "s": "0x42e19d62c96d87683833bb0a02150c12580e61c5d69f6c17753cab8e3c726208",
      "status": "success",
      "to": "0xD216153c06E857cD7f72665E0aF1d7D82172F494",
      "toInfo": Object {
        "contractName": undefined,
        "isContract": true,
      },
      "transactionIndex": 0,
      "v": "0x1c",
      "value": 0,
    },
    "time": 1581720675000,
    "type": "transaction",
  },
  Object {
    "body": Object {
      "address": "0xBb2692EC9976975A8cB80B52832A94273fBFB438",
      "addressInfo": Object {
        "contractName": undefined,
        "isContract": true,
      },
      "blockHash": "0x05a82978c95cf9e30045e0eb9b9cb48d5639b3ab9b8007b2f8d371dc99ed0afd",
      "blockNumber": 8394957,
      "data": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "event": Object {
        "args": undefined,
        "name": "Sent",
        "params": Array [
          Object {
            "name": undefined,
            "type": "address",
            "value": "0xBE7208Ede01673d0b7338fEf14cdd9218501ab81",
          },
          Object {
            "name": undefined,
            "type": "address",
            "value": "0xBE7208Ede01673d0b7338fEf14cdd9218501ab81",
          },
          Object {
            "name": undefined,
            "type": "address",
            "value": "0x0f5eF980513cA22fAF70243959c92cf5eE1E531B",
          },
          Object {
            "name": undefined,
            "type": "uint256",
            "value": 0,
          },
          Object {
            "name": undefined,
            "type": "bytes",
            "value": null,
          },
          Object {
            "name": undefined,
            "type": "bytes",
            "value": null,
          },
        ],
        "signature": "Sent(address,address,address,uint256,bytes,bytes)",
      },
      "logIndex": 0,
      "removed": false,
      "topics": Array [
        "0x06b541ddaa720db2b10a4d0cdac39b8d360425fc073085fac19bc82614677987",
        "0x000000000000000000000000be7208ede01673d0b7338fef14cdd9218501ab81",
        "0x000000000000000000000000be7208ede01673d0b7338fef14cdd9218501ab81",
        "0x0000000000000000000000000f5ef980513ca22faf70243959c92cf5ee1e531b",
      ],
      "transactionHash": "0xff36619285cf23ef7f812ea3600a369b52ebd392e682deeb77dd758b4462f38f",
      "transactionIndex": 0,
    },
    "time": 1581720675000,
    "type": "event",
  },
  Object {
    "body": Object {
      "address": "0xBb2692EC9976975A8cB80B52832A94273fBFB438",
      "addressInfo": Object {
        "contractName": undefined,
        "isContract": true,
      },
      "blockHash": "0x05a82978c95cf9e30045e0eb9b9cb48d5639b3ab9b8007b2f8d371dc99ed0afd",
      "blockNumber": 8394957,
      "data": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "event": Object {
        "args": undefined,
        "name": "Transfer",
        "params": Array [
          Object {
            "name": undefined,
            "type": "address",
            "value": "0xBE7208Ede01673d0b7338fEf14cdd9218501ab81",
          },
          Object {
            "name": undefined,
            "type": "address",
            "value": "0x0f5eF980513cA22fAF70243959c92cf5eE1E531B",
          },
          Object {
            "name": undefined,
            "type": "uint256",
            "value": 0,
          },
        ],
        "signature": "Transfer(address,address,uint256)",
      },
      "logIndex": 1,
      "removed": false,
      "topics": Array [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x000000000000000000000000be7208ede01673d0b7338fef14cdd9218501ab81",
        "0x0000000000000000000000000f5ef980513ca22faf70243959c92cf5ee1e531b",
      ],
      "transactionHash": "0xff36619285cf23ef7f812ea3600a369b52ebd392e682deeb77dd758b4462f38f",
      "transactionIndex": 0,
    },
    "time": 1581720675000,
    "type": "event",
  },
  Object {
    "body": Object {
      "address": "0xD216153c06E857cD7f72665E0aF1d7D82172F494",
      "addressInfo": Object {
        "contractName": undefined,
        "isContract": true,
      },
      "blockHash": "0x05a82978c95cf9e30045e0eb9b9cb48d5639b3ab9b8007b2f8d371dc99ed0afd",
      "blockNumber": 8394957,
      "data": "0xd80106b900000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000015ccba569c700",
      "event": Object {
        "args": undefined,
        "name": "TransactionRelayed",
        "params": Array [
          Object {
            "name": undefined,
            "type": "address",
            "value": "0x9718A022941447fa2aF776eCbde7007bBA5A7edd",
          },
          Object {
            "name": undefined,
            "type": "address",
            "value": "0x150042AAe387E5c002dfaF7844b3110828094ACA",
          },
          Object {
            "name": undefined,
            "type": "address",
            "value": "0x6C85E15838034B7BbA6C63bAdD8839fd4c5BE606",
          },
          Object {
            "name": undefined,
            "type": "bytes4",
            "value": "0xd80106b9",
          },
          Object {
            "name": undefined,
            "type": "uint8",
            "value": 0,
          },
          Object {
            "name": undefined,
            "type": "uint256",
            "value": 383504700000000,
          },
        ],
        "signature": "TransactionRelayed(address,address,address,bytes4,uint8,uint256)",
      },
      "logIndex": 2,
      "removed": false,
      "topics": Array [
        "0xab74390d395916d9e0006298d47938a5def5d367054dcca78fa6ec84381f3f22",
        "0x0000000000000000000000009718a022941447fa2af776ecbde7007bba5a7edd",
        "0x000000000000000000000000150042aae387e5c002dfaf7844b3110828094aca",
        "0x0000000000000000000000006c85e15838034b7bba6c63badd8839fd4c5be606",
      ],
      "transactionHash": "0xff36619285cf23ef7f812ea3600a369b52ebd392e682deeb77dd758b4462f38f",
      "transactionIndex": 0,
    },
    "time": 1581720675000,
    "type": "event",
  },
]
`;
